kafka-connect-aws-s3/src/test/scala/io/lenses/streamreactor/connect/aws/s3/formats/BytesFormatWriterTest.scala

```diff
@@ -55,6 +58,115 @@ class BytesFormatWriterTest extends AnyFlatSpec with Matchers {
+    val topic        = Topic("myTopic")
+    data.foreach {
+      case value =>
+        val v: SinkData = value.fold(NullSinkData(None).asInstanceOf[SinkData])(v =>
+          ByteArraySinkData(v.getBytes(), None).asInstanceOf[SinkData],
+        )
+        writer.write(None, v, topic)
+    }
+
+    val result = outputStream.toByteArray
+
+    val reader = new BytesFormatWithSizesStreamReader(
+      new ByteArrayInputStream(result),
+      result.length.toLong,
+      mock[S3Location],
+      BytesWriteMode.ValueWithSize,
+    )
+    data.foreach {
+      case value =>
+        withClue(s"value: $value") {
+          val byteArraySourceData = reader.next()
+          Option(byteArraySourceData.data.key).filter(_.nonEmpty) shouldBe None
+          Option(byteArraySourceData.data.value).filter(_.nonEmpty).map(new String(_)) shouldBe value
+        }
+    }
+  }
+
+  "Keys with Size" should "round trip" in {
+    val data = List(
+      "record 1 value".some,
+      "record 2 value".some,
+      "record 3 value".some,
+      None,
+    )
+    val outputStream = new S3ByteArrayOutputStream()
+    val writer       = new BytesFormatWriter(outputStream, BytesWriteMode.KeyWithSize)
+    val topic        = Topic("myTopic")
+    data.foreach {
+      case value =>
+        val v: SinkData = value.fold(NullSinkData(None).asInstanceOf[SinkData])(v =>
+          ByteArraySinkData(v.getBytes(), None).asInstanceOf[SinkData],
+        )
+        writer.write(v.some, NullSinkData(None), topic)
+    }
+
+    val result = outputStream.toByteArray
+
+    val reader = new BytesFormatWithSizesStreamReader(
+      new ByteArrayInputStream(result),
+      result.length.toLong,
+      mock[S3Location],
+      BytesWriteMode.KeyWithSize,
+    )
+    data.foreach {
+      case value =>
+        withClue(s"key: $value") {
+          val byteArraySourceData = reader.next()
+          Option(byteArraySourceData.data.value).filter(_.nonEmpty) shouldBe None
+          Option(byteArraySourceData.data.key).filter(_.nonEmpty).map(new String(_)) shouldBe value
+        }
+    }
+  }
+
+  "Values and Keys" should "round trip" in {
+    val data: List[(Option[String], Option[String])] = List(
+      Option.empty[String] -> "record 1 value".some,
+      "record 2 key".some  -> "record 2 value".some,
+      None                 -> "record 3 value".some,
+      "record 4 key".some  -> None,
+    )
+    val outputStream = new S3ByteArrayOutputStream()
+    val writer       = new BytesFormatWriter(outputStream, BytesWriteMode.KeyAndValueWithSizes)
+    val topic        = Topic("myTopic")
+    data.foreach {
+      case (key, value) =>
+        val k: Option[SinkData] = key.map(v => ByteArraySinkData(v.getBytes()))
+        val v: SinkData = value.fold(NullSinkData(None).asInstanceOf[SinkData])(v =>
+          ByteArraySinkData(v.getBytes(), None).asInstanceOf[SinkData],
+        )
+        writer.write(k, v, topic)
+    }
```

Comment:
Again very similar, could some common code be extracted from the 3 similar blocks?